minetest/minetest
# Dehardcode the death formspec

Fixes #14218. Similar to #8014, but with full backwards compatibility. 

The death screen is now part of builtin and can be overriden by replacing `minetest.show_death_screen`. It's sent as a normal formspec. The server no longer sends TOCLIENT_DEATHSCREEN or listens for TOSERVER_RESPAWN. 

Backwards compatibility:

- New server + old client: No special handling required, same as new server + new client. The server doesn't send TOCLIENT_DEATHSCREEN, so the client doesn't show the hardcoded death screen. The server sends a normal formspec as the death screen.

- Old server + new client: The client still listens for TOCLIENT_DEATHSCREEN, shows the hardcoded death screen and sends TOSERVER_RESPAWN when necessary.

## To do

This PR is a Ready for Review.

## How to test

```lua

local old_show_death_screen = minetest.show_death_screen
minetest.show_death_screen = function(player, reason)
	if reason and reason.type == "node_damage" and
			minetest.get_item_group(reason.node, "igniter") ~= 0 then
		minetest.chat_send_player(player:get_player_name(), "You have been fried.")
		player:respawn()
		return
	end
	old_show_death_screen(player, reason)
end
```

Confirm that this works for new and old clients alike.

Confirm that you still see a death screen when joining old servers.
